### ISSUE TITLE
download without --from?

### ISSUE THREAD

**Issue by schoettl** [dpaetzel/aq2ledger#2]
Just a question (I cannot yet test it):

What's the behavior when calling `aq2ledger download` without `--from`?

And does it recognize and merge duplicate entries?

Thanks!

**Comment by dpaetzel** [comment]
Currently, `--from` is required; I myself have all the data up to some pre-PSD2 date and don't need to re-download anything prior to that date.

Besides that, it seems that some banks (e.g. DKB) restrict downloading transactions to a certain number of days (e.g. 160) since PSD2. I'm currently trying to solve that by automatedly starting several downloads of a certain number of days (e.g. 100). This is especially simple to do, if the first date to be downloaded is specified by the user as otherwise, we would need to download and then decide what to do after inspecting the downloaded data. Since different banks do stuff very differently, I'm not sure whether that would be a stable approach.

Note that `--from` for `print` is mandatory as well as there is no merging of journals (yet).

**Comment by schoettl** [comment]
> Currently, `--from` is required; I myself have all the data up to some pre-PSD2 date and don't need to re-download anything prior to that date.

OK, thank you for the information.

Hm, I don't want to define a `--from` date. I just want to download and add all new transactions automatically. This is why I use buchhaltung. So in my opinion, *merging* / *ignoring duplicates*  would be an important feature.

**Comment by dpaetzel** [comment]
> Hm, I don't want to define a --from date. I just want to download and add all new transactions automatically. This is why I use buchhaltung. So in my opinion, merging / ignoring duplicates would be an important feature.

I agree, `buchhaltung`'s merging of journals was very handy. Note that `aq2ledger` is still work in progress and merging is one of the features I plan to implement next (see the README). For me, it's not that big a deal to quickly check the latest date in my journal file and then use that for `--from` which is why I thought I'd release the current state albeit it being incomplete.

**Comment by dpaetzel** [comment]
> This is why I use buchhaltung.

Does that mean that `buchhaltung` still works for you, even after PSD2?

**Comment by schoettl** [comment]
> For me, it's not that big a deal to quickly check the latest date in my journal file and then use that for `--from` which is why I thought I'd release the current state albeit it being incomplete.

I fear that doing it this way, it's possible to miss some transactions or to get duplicates.

E.g., if your last download was on some day at 12:00 noon and there are transactions before *and* after noon.

> Does that mean that buchhaltung still works for you, even after PSD2?

Yes, buchhaltung works for me. It did not work for a while. But after some updates of aqbanking it worked again. So I think, only aqbanking together with my configuration had some problems after PSD2.

**Comment by dpaetzel** [comment]
> I fear that doing it this way, it's possible to miss some transactions or to get duplicates.

I'm aware that I get duplicates for a single day with high probability that way (which is OK for me, personally). But as I said, merging will be implemented :slightly_smiling_face: 

> Yes, buchhaltung works for me.

Oh, OK, I wasn't aware that it works again :smile: Do you use my fork or the original repository? `buchhaltung` working again means that I might have done some unnecessary work :thinking: :smile:

**Comment by schoettl** [comment]
Oh, I forgot that I currently use your fork of buchhaltung. And that fork works for me for the last few weeks.

I still like the "do one thing and do it well" philosophy of your project.

**Comment by dpaetzel** [comment]
> that fork works for me for the last few weeks.

Nice!

Does `buchhaltung` still download all the transactions ever occurred for you? For one of my banks, `aqbanking-cli` (used by `buchhaltung`) does only download transactions from time windows of different sizes depending on which `--fromdate` I use … For some values of `--fromdate` it's 160 days, for some just a single month (and if I leave out `--fromdate` it doesn't download all transactions ever occurred either–as it did pre-PSD2). It's probably just some strange handling of PSD2 by one of my banks but who knows?

**Comment by schoettl** [comment]
Actually, I don't know. I have to input a TAN every time I update, so I guess it downloads more than just the last few weeks.

**Comment by dpaetzel** [comment]
You're right. I finally found out what happens: DKB seems to only ever give you ~100 (sometimes 100, sometimes 103, sometimes 101) transactions at once and they do not even give you all the transactions of the last day they give you transactions for. This means that I will have to change `aq2banking`'s workflow fundamentally (parsing cannot be separated from downloading to `.ctx`-files any more because we have to parse the downloaded transactions and check for whether everything we wanted is actually there or whether we need to internally adjust the `--fromdate` given to `aqbanking-cli` and start another download).

`buchhaltung` doesn't do that either afaik, so you might be missing transactions, depending on which banks you have accounts at.
